libraries/scripting/type-providers/src/kotlin/script/experimental/typeProviders/generatedCode/GeneratedCode.kt

```diff
@@ -0,0 +1,302 @@
+                mutableCodeList.addAll(codeList)
+            } else {
+                mutableCodeList.add(this)
+            }
+        }
+
+        operator fun String.unaryPlus() {
+            mutableCodeList.add(inlineCode(this))
+        }
+
+        internal fun build(): GeneratedCode {
+            return composite(mutableCodeList)
+        }
+    }
+
+    object Empty : GeneratedCode by EmptyCode()
+
+    companion object {
+        /**
+         * Create a composition of multiple instances of Generated Code
+         */
+        fun composite(init: Builder.() -> Unit) = Builder().apply(init).build()
+
+        /**
+         * Create a composition of multiple instances of Generated Code
+         */
+        fun composite(iterable: Collection<GeneratedCode>): GeneratedCode = if (iterable.count() == 1)
+            iterable.first()
+        else
+            CompoundGeneratedCode(iterable.toList())
+
+        /**
+         * Create a composition of multiple instances of Generated Code
+         */
+        fun composite(vararg code: GeneratedCode) = composite(code.toList())
+    }
+}
+
+private data class CompoundGeneratedCode(val codeList: List<GeneratedCode>) : InternalGeneratedCode() {
+    override fun StringBuilder.write(indent: Int) {
+        writeJoined(codeList, "\n\n", indent)
+    }
+
+    override fun GeneratedCodeVisitor.visit() {
+        for (code in codeList) {
+            visit(code)
+        }
+    }
+}
+
+/**
+ * References a member that can be used as a type
+ */
+interface IdentifiableMember {
+    val name: String
+    fun imports(): Set<String> = emptySet()
+
+    companion object {
+        private data class KClassIdentifiable(val kClass: KClass<*>) : IdentifiableMember {
+            override val name: String = kClass.simpleName ?: kClass.qualifiedName!!
+            override fun imports() = setOfNotNull(kClass.qualifiedName)
+        }
+
+        private data class NamedIdentifiable(override val name: String) : IdentifiableMember
+
+        operator fun invoke(kClass: KClass<*>): IdentifiableMember = KClassIdentifiable(kClass)
+        operator fun invoke(name: String): IdentifiableMember = NamedIdentifiable(name)
+    }
+}
+
+/**
+ * References an interface that can be implemented
+ */
+interface GeneratedInterface : IdentifiableMember
+
+private data class CastedInterfaceMember(private val member: IdentifiableMember) : GeneratedInterface, IdentifiableMember by member
+
+fun IdentifiableMember.asInterface(): GeneratedInterface = CastedInterfaceMember(this)
+
+/**
+ * Generated code that is a type that can be used as an Argument
+ */
+interface IdentifiableCode : GeneratedCode, IdentifiableMember
+
+/**
+ * Builder for a provided type whose JVM Name can be set
+ */
+interface JVMNameableBuilder {
+    fun jvmName(name: String)
+}
+
+/**
+ * Builder for a provided type. It can implement an interface and have some code inside.
+ */
+interface GeneratedTypeBuilder : IdentifiableMember {
```

Comment:
This seems to represent an analog of `ClassLike` things in the compiler, so maybe it makes sense to name it accordingly. `Type` is a bit more general.

Reply:
Renamed to ClassLike